### PR TITLE
add reparameterized property to distributions

### PR DESCRIPTION
**Pull request by alicanb**
add `.reparametrized` property to distributions
- add `__init__` function to distributions (currently only assigns `.reparametrized` property)
- add tests to check outputs of `Normal.sample` when reparametrized
- addresses https://github.com/probtorch/pytorch/issues/7

### CODE REVIEW

**Comment by jwvdm** [comment]
I have a tiny procedural comment: Can we decide on which way we will spell reparameterized?

1. reparametrized (this is what @alicanb has)
2. reparameterized (this is what Tensorflow has)

I don't mind either way, as long as we enforce consistency from here on out :)

**Comment by jwvdm** [comment]
(copied this comment in #7)

File: torch/distributions.py [hunk 1]
```diff
@@ -41,7 +41,9 @@ class Distribution(object):
     r"""
     Distribution is the abstract base class for probability distributions.
     """
-
+    def __init__(self, reparameterized = False):
+        self.reparameterized = reparameterized
```
Comment (fritzo) [inline torch/distributions.py]: I think this should be a per-class property. You can eliminate the `__init__` method and simply set
```py
class Distribution(object):
    reparameterized = False
    ...
class Normal(Distribution):
    reparameterized = True
    ...
```

File: torch/distributions.py [hunk 2]
```diff
@@ -168,14 +174,20 @@ class Normal(Distribution):
...
 
-    def __init__(self, mean, std):
+    def __init__(self, mean, std, reparameterized = False):
         self.mean = mean
         self.std = std
+        super(Normal, self).__init__(reparameterized=reparameterized)
 
     def sample(self):
-        return torch.normal(self.mean, self.std)
+        if self.reparameterized:
```
Comment (fritzo) [inline torch/distributions.py]: As I understand `.reparameterized` is merely an "under construction" marker that says "if you call `your_dist.sample(...)` with `Variable`s that require gradients, then you'll get a `NotImplementedError`". It looks like you're using it as an alternative to `mean.requires_grad` and `std.requires_grad`. I think this conditional should simply be:
```py
if self.mean.requires_grad or self.std.requires_grad:
    eps = torch.normal(torch.zeros_like(self.mean), torch.ones_like(self.std))
    return self.mean + self.std * eps
else:
    return torch.normal(self.mean, self.std)
```
What do you think?

File: torch/distributions.py [hunk 3]
```diff
@@ -168,14 +174,20 @@ class Normal(Distribution):
     Args:
         mean (float or Tensor or Variable): mean of the distribution
         std (float or Tensor or Variable): standard deviation of the distribution
+        reparameterized (bool): whether to use reparameterization trick
```
Comment (fritzo) [inline torch/distributions.py]: The `reparameterized` docs should probably live a `Properties:` or `Attributes:` section.

**Comment by fritzo** [comment]
What do you think of adding a `requires_grad` argument to `.sample()` methods?
```py
class Normal(distribution):
    reparameterized = True  # requires_grad is fully implemented
    def sample(requires_grad=False):
        if not requires_grad:
            return torch.normal(self.mean, self.std)
        eps = torch.normal(torch.zeros_like(self.mean), torch.ones_like(self.std))
        return self.mean + self.std * eps
class Gamma(distribution):
    reparameterized = False  # requires_grad is only partially implemented
    def sample(requires_grad=False):
        if not requires_grad:
            return torch.random_gamma(self.alpha, self.beta)
        if not self.alpha.requires_grad:
            # easy special case
            return torch.random_gamma(self.alpha, 1.0) / self.beta
        raise NotImplementedError('https://github.com/pytorch/pytorch/issues/3813')
```

**Comment by alicanb** [comment]
This is the closest to what I have been thinking about. Essentially, only
sampling functions use reparameterization property(as requires_grad) , I
guess there is no need to expose it to the rest of the implementation. In
this way, is there further need for keeping the reparameterization as a
property of the distribution?

**Comment by fritzo** [comment]
> is there further need for keeping the reparameterization as a property of the distribution?

I'm leaning towards omitting `.reparameterized` in favor if `.is_discrete` and then implementing reparameterized versions of all continuous distributions. @srush do you have an opinion?

**Comment by jwvdm** [comment]
@fritzo: I would say that `.reparamterized` is probably more accurate than `.is_discrete`, unless we are convinced that we will never ship a distribution that is continuous yet not reparameterized (see also #7).

**Comment by alicanb** [comment]
@fritzo can you check it one more time before I push it upstream?

File: test/test_distributions.py [hunk 4]
```diff
@@ -75,11 +85,40 @@ def test_normal(self):
         self.assertEqual(Normal(mean_1d, std_1d).sample().size(), (1,))
         self.assertEqual(Normal(0.2, .6).sample_n(1).size(), (1, 1))
         self.assertEqual(Normal(-0.7, 50.0).sample_n(1).size(), (1, 1))
+
```
Comment (fritzo) [inline test/test_distributions.py]: nit: We should eventually split these tests up into smaller pieces, but that can wait.
